data/fwupd-offline-update.service.in

```diff
@@ -11,3 +11,4 @@ Before=shutdown.target system-update.target
 Type=oneshot
 ExecStart=@libexecdir@/fwupd/fwupdoffline
 FailureAction=reboot
+FailureActionExitStatus=0
```

Comment:
So we rewrite the exit status to 0 on failure? Not sure I understood the remove-reboot-exit-failure logic...

Reply:
Let me try to explain it better.
1. In the situation that the offline target isn't intended to be fwupd, return `EXIT_SUCCESS`, but no reboot so that whatever the actual target is can run.
2. In the situation that fwupd is the offline target, need to reboot either way at the end.
3. If you call reboot before the application exits, then it gets `SIGTERM`


Do you think it would be clearer if it was inverted?
* So the calling fwupd-offline-update when something else is offline target, you return `EXIT_FAILURE`.
* In all other cases return `EXIT_SUCCESS`
* Adjust the systemd unit to reboot on `SuccessAction` instead of `FailureAction`

Reply:
tbh, it all seems very complicated. I wonder if we should just not use the offline mode at all -- it's even causing problems on OSTree systems as `/` is immutable. What about if fwupdoffline was just run it with `WantedBy=final.target` and we did the upgrade on shutdown rather than with two new reboots? Two birds one stone.

The super important bit for me is the forced restart of the keyboard controller after flashing (as the fans might not work until we do!) but I think I can reset any EC parameters set before flashing. Before I test if this is even possible, I guess this is an order of magnitude easier to implement for the snap?

Reply:
> What about if fwupdoffline was just run it with WantedBy=final.target and we did the upgrade on shutdown rather than with two new reboots? Two birds one stone.

If it was moved to shutdown then I sort of think we may as well just combine the activate and offline job.  In many ways they're similar, calling into a device's vfunc to do some action to the device.

I do like that for offline updates it was essentially `ln -s /system-update $SOMEWHERE` and it only runs that target during that situation.  I think it can be adopted still.  So how about something like this in the systemd job that would run on shutdown:

`ConditionPathExists=@localstatedir@/lib/fwupd/offline-pending`

And the the first thing the daemon does next time it's started is removes that file?

>  Before I test if this is even possible, I guess this is an order of magnitude easier to implement for the snap?

Yes, I do think so.

Reply:
> we may as well just combine the activate and offline job

Not quite; the shutdown job would be run with /usr still mounted and the activation would be run with /usr unmounted.

> ConditionPathExists=@localstatedir@/lib/fwupd/offline-pending

Yes, I think that makes sense. Give me an hour or so.

Reply:
Ideas please: How could the daemon "know" it's running in offline mode, i.e. before `shutdown.target`? We can't do `{GetUnit}.Get(ActiveState)` as the unit is not (yet) on the bus when the D-Bus method is called. I really want to avoid the client saying to the daemon "I'm offline now, TRUST ME" and work it out in the daemon.

Reply:
Actually, just a bug, no race.

Reply:
I've pushed a branch wip/hughsie/offline-at-shutdown with some ideas. Zero testing, but I thought it might be a good way to start thinking about how to do it properly at shutdown. I think you two understand the systemd unit more than me.